### PR TITLE
7124: Add focus and maximizedFocus launch modes

### LINKED ISSUE #7124: Focus mode on app startup/launch
The latest terminal preview:  1.2.2022.0 has "togglefocusmode" option which is great. Can we have a setting that starts the terminal in this mode by default? I want to have maximized and focused mode when app is launch

Or is this already available? I couldn't find any links/references to this

### LINKED ISSUE #7825: focusMode option for launchMode setting
# Description of the new feature/enhancement
Besides having maximized and fullscreen, would it be possible to please also have the option to start the terminal in FocusMode?

# Proposed technical implementation details (optional)

### LINKED ISSUE #7875: Add command line flag to start in focus mode
# Description of the new feature/enhancement

#6804 made it possible to strip the terminal of window decorations with the `toggleFocusMode` command.

It would be nice to invoke the terminal with this behavior already applied.

# Proposed technical implementation details

A simple command line flag like:

```
wt.exe --focus-mode
```

to simplify the process:

```
wt.exe
*wait for terminal window to open*
*open command palette or press keybinding to start focus mode*
```

### PR DESCRIPTION
**Pull request by Don-Vito**
This commit introduces two new launch modes: focus and maximizedFocus. 
* Focused mode, behaves like a default mode, but with the Focus Mode
  enabled.
* Maximized focused mode, behaves like a Maximized mode, but with the
  Focus Mode enabled.

There two ways to invoke these new modes:
* In the settings file: you set the "launchMode" to either "focus" or
  "maximizedFocus"
* In the command line options, you can path -f / --focus, which is
  mutually exclusive with the --fullscreen, but can be combined with the
  --maximized:
  * Passing -f / --focus will launch the terminal in the "focus" mode
  * Passing -fM / --focus --maximized will launch the terminal in the
    "maximizedFocus" mode

This should resolve a relevant part in the command line arguments
mega-thread #4632

Closes #7124
Closes #7825
Closes #7875

### CODE REVIEW

**Comment by zadjii-msft** [comment]
I love this. I think you've done the `focused` mode thing exactly right. My one concern - I'd yank the `maximizedFocus` mode. In my opinion, that's just the same thing as fullscreen mode 😄

**Comment by Don-Vito** [comment]
> 
> 
> I love this. I think you've done the `focused` mode thing exactly right. My one concern - I'd yank the `maximizedFocus` mode. In my opinion, that's just the same thing as fullscreen mode 😄

The only difference from the fullscreen is that in the maximized focused mode, you still see the taskbar (not sure how useful it is). Added this mode as it seems that in was originally mentioned in the feature request (https://github.com/microsoft/terminal/issues/7124): "I want to have maximized and focused mode when app is launch" (https://github.com/microsoft/terminal/issues/7124#issue-668830767)

Please let me know if you prefer me to delete anyway.

**Comment by zadjii-msft** [comment]
Oh, fair point! Then I'm totally cool with this. I don't hate `maximizedFocus` either - I think `maximized|focus` would be wrong, because that would imply the whole set is flags that could be combined (they aren't). It's basically either `maximizedFocus` or `focusMaximized`, and I don't feel like there's a tangible difference between the two, so let's go with what's already been implemented 😛

**Comment by Don-Vito** [comment]
> 
> 
> Oh, fair point! Then I'm totally cool with this. I don't hate `maximizedFocus` either - I think `maximized|focus` would be wrong, because that would imply the whole set is flags that could be combined (they aren't). It's basically either `maximizedFocus` or `focusMaximized`, and I don't feel like there's a tangible difference between the two, so let's go with what's already been implemented 😛

Yay :) 
Should I add some UTs? If so - can you please guide me to the framework that I can base on? 
If not - should I mark this as "Ready for review"? (I am quite new with the Draft PRs, work with Azure DevOps mostly).

**Comment by zadjii-msft** [comment]
> Should I add some UTs? 

Bless your soul ☺️ 

In this specific case, since this is such a UI-dependent feature, I think we're cool not adding any unit tests. The enum parsing is already pretty well tested, and I'm not sure we've got any sort of framework for testing the state of the Win32 window. I'd definitely mark this as "Ready for Review" 😄

File: src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsModel/TerminalSettingsSerializationHelpers.h:172]: Man this is gonna seem like the most trivial nit, but maybe the string should just be `focus`, as in "I want the launch mode to be focus mode". I think if it's `focused`, then that makes me think "I want the terminal to launch and be focused", implying there's an `unfocused` for opening the terminal w/o focusing the window (which is _not_ something we plan on adding, certainly not in this property).

(& also change the below to `maximizedFocus`)

**Review by zadjii-msft — APPROVED** [review]

**Comment by Don-Vito** [comment]
@zadjii-msft -  sorry for pushing another change after  the approval, I have noticed that in the megathread (https://github.com/microsoft/terminal/issues/4632) there is also a request to add command-line option for this. So added it as well.

**Review by zadjii-msft — APPROVED** [review]
_are you a mind reader?_

I was gonna go ahead and sneak the cmdline argument in after this PR merged, but hey, you've gone and already done it. Thanks for the diligence! I quite like the elegance of `-Mf` for combining focus and maximized too

**Review by DHowett — CHANGES_REQUESTED** [review]
I believe we'll need to update the schema in doc/cascadia! This looks excellent otherwise.

**Comment by Don-Vito** [comment]
Added the schema to the cascadia docs. Should I do anything beyond validating that it works on different profile? 
This schema corresponds to the documentation PR I put in the docs repo (https://github.com/MicrosoftDocs/terminal/pull/158)

File: src/cascadia/TerminalApp/AppCommandlineArgs.cpp
Comment (lhecker) [inline src/cascadia/TerminalApp/AppCommandlineArgs.cpp:178]: You can drop the `winrt::Microsoft::Terminal::Settings::Model::` prefix, because the file [already imports the namespace](https://github.com/microsoft/terminal/blob/ad03ae0ab889f81b76ed339619e739cd5c9c468a/src/cascadia/TerminalApp/AppCommandlineArgs.cpp#L10). 👍

**Review by lhecker — COMMENTED** [review]
I believe you can drop `MaximizedFocusMode`.
You could get the same effect by calling `wt.exe --focus --maximized` instead, couldn't you?

**Comment by Don-Vito** [comment]
> 
> 
> I believe you can drop `MaximizedFocusMode`.
> You could get the same effect by calling `wt.exe --focus --maximized` instead, couldn't you?

@lhecker - please guide me here. How do you suggest to encode the combination of --focus --maximized instead?

File: src/cascadia/TerminalApp/AppCommandlineArgs.cpp
Comment (Don-Vito) [reply src/cascadia/TerminalApp/AppCommandlineArgs.cpp]: Thanks! Done!

**Comment by Don-Vito** [comment]
@DHowett - sorry for nagging, but CI is somewhat harsh with me today and I am not sure if I can somehow unblock this PR 😄 
* The github spelling complains about something I already fixed (I wrote betwen instead of between). 
* The OutputTests::BasicReadConsoleOutputATest fails - though it passes locally for me, and I can hardly follow the link between it and the change.

**Review by DHowett — APPROVED** [review]
Excellent, excellent stuff. Thank you!

**Comment by DHowett** [comment]
>## I have several questions regarding the feature and the PR

>* This is my first PR, please be gentle 

This is excellent work for a first PR -- though your other work merged first! :smile:
We're happy to have you.

>* I hesitated between introducing new launch modes or adding a separate setting for "startInFocused" mode. I >preferred not to introduce a new parameter, was this a good decision?

Looks like we went w/ a parameter.

>* Didn't updated the documentation yet - wanted to hear your opinion first.

All set!

>* Not sure how to test this in UT. Didn't find UT for other modes.

@zadjii-msft summed it up quite nicely.

>* Didn't add those modes to args. Please let me know if required (and if so, the names of the options)

All set again! 😁

**Pull request merged**
